### PR TITLE
SBOL UML first implementation

### PR DESCRIPTION
**Pull request by jakebeal**
UML implementation working as far as smoke tests of library creation and simple calibration protocol

### CODE REVIEW

**Review by danbryce — APPROVED** [review]
I didn't go over the UML modeling in great detail, but I reviewed the LUDOX protocol representation.  It has what I need to extract the activity relationships.  

I'm not sure why you chose to comment out lots of code instead of delete it, but that might be one thing to improve.  

I was able to run all the tests and have them pass.

**Comment by jakebeal** [comment]
Thank you, @danbryce ; the commented code is next on the block for updates, and just needed to be out of the way to keep from interfering in current runs. Not the cleanest practice, but it shouldn't last long.

**Comment by danbryce** [comment]
My other observation is that you are keeping a list of behaviors and treating the protocol as sequential (at least in the example I saw).  I think that a list is fine, but it may be better as a set.  The flows will imply a partial ordering that may be different from the total ordering implied by the list.

**Comment by jakebeal** [comment]
The representation and library support that: just use `execute_primitive` instead of `primitive_step`.
The `primitive_step` function is a macro for `execute_primitive` plus `order(last_step,this_step)`, to allow easier translation of "paper protocols".

**Review by rpgoldman — COMMENTED** [review]
Wouldn't it be more normal to have `__init__.py` only have import statements? I was kind of surprised to see "real code" in here.

**Comment by rpgoldman** [comment]
Sorry to be so dense, but what do you mean by "UML implementation"? Is it that `uml.ttl` is the core of this PR?  I'm trying to find my way through this PR, but I need a little bit more explanation in PR description; I don't know what's being attempted, so it's hard to know how to evaluate it.

**Comment by jakebeal** [comment]
The code in `__init__` is an sbol_factory pattern. @bbartley and I are discussing potential ways to make this more "normal", but right now that's the monkey-patching pattern is the only way we've figured out for adding methods to the autogenerated classes.  Potential alternatives under consideration:
 - Generating code and inserting chunks
 - Generating code that then respects blocks edited with a certain format
 - Inserting "silent" extension classes into the inheritance hierarchy
 
None of these seem significantly less ugly; if you can figure out a better alternative, I'm happy to hear it.

**Comment by jakebeal** [comment]
@rpgoldman 

> Sorry to be so dense, but what do you mean by "UML implementation"? Is it that `uml.ttl` is the core of this PR? I'm trying to find my way through this PR, but I need a little bit more explanation in PR description; I don't know what's being attempted, so it's hard to know how to evaluate it.

You are correct. `uml.ttl` is the core of the PR, with `uml/__init__.py` the helper classes to allow one to build protocols compactly with it.  Everything else is either consequences of that changes or tests validating that it lets us specify a protocol.

**Comment by rpgoldman** [comment]
One possible issue: the `uml.ttl` file seems to use Turtle* notation, rather than simple Turtle. The difference comes when dealing with quads and reification.  I ran the following command to translate `uml.ttl` to `uml-flattened.ttl`:

```
docker run -v `pwd`:/models turtlestar2turtle /models/uml.ttl > uml-flattened.ttl
```

I can't figure out how to attach the file to the issue, so I have committed it. If it's important to have a translation, we will have to figure out how to keep it synced and also figure out how tools that won't process Turtle* can download the flattened equivalent.

File: uml/uml.ttl
Comment (rpgoldman) [inline uml/uml.ttl:82]: I believe that UML makes the type attribute optional ( 0..1 ), but do we want to? Should we just make the cardinality one and refuse to let people leave types unspecified?

Your call.

File: uml/uml.ttl
Comment (rpgoldman) [inline uml/uml.ttl:49]: Am I right in thinking that `MultiplicityElement` has no existence in the ontology proper?  Can it appear as a URI in the `uml:type` property?  If so, should we be declaring its existence as an entity?

I'm not sure I properly understand what is and is not meta here.

**Review by rpgoldman — APPROVED** [review]
I have some thoughts about the OWL issues here (but not really as an OWL expert, just twice burned, four times careful).  I'm somewhat uncomfortable with the fact that the limitations on OWL mean that a lot of the content of the formalism here is relatively invisible to the logics of RDFS and OWL.  But that probably says more about the limitations of those notations than what you have drawn up here.
I'm afraid I don't have any useful input on UML, which I have never used. 
As a troll, I just think of UML diagrams as those things which someone writes at the start of a project, and which then get progressively out of date as the project proceeds. Typically they are drawn by an "object modeler" who thinks deeply about the domain, but does not think at all about what needs to be computed.
I understand that using UML is The Right Thing from the standpoint of standards compliance, but we should get a review from someone who understands UML much better than I do.

File: uml/uml.ttl
Comment (jakebeal) [reply uml/uml.ttl:82]: Leaving it unspecified allows the potential for type inference to be used where possible, which I'd like to not prohibit.

File: uml/uml.ttl
Comment (jakebeal) [reply uml/uml.ttl:78]: Added an issue for this as #28

File: uml/uml.ttl
Comment (jakebeal) [reply uml/uml.ttl:49]: TypedElement and MultiplicityElement are abstract interfaces that require multiple inheritance that sbol_factory isn't set up to support at present.

**Pull request merged**
